File: 04.2_diffie_hellman/DiffieHellman.java
//To Run this code use command-line and cd into this folder and Run: javac DiffieHellman.java && java DiffieHellman

import java.math.BigInteger;

public class DiffieHellman {

    // Power function to return value of a^b mod P
    public static long power(long a, long b, long P) {
        BigInteger A = BigInteger.valueOf(a);
        BigInteger B = BigInteger.valueOf(b);
        BigInteger Pmod = BigInteger.valueOf(P);

        BigInteger result = A.modPow(B, Pmod);
        return result.longValue();
    }

    // Encrypt/Decrypt function using XOR
    public static String encryptDecrypt(String message, long key) {
        StringBuilder result = new StringBuilder();
        for (char c : message.toCharArray()) {
            result.append((char) (c ^ key));
        }
        return result.toString();
    }

    public static void main(String[] args) {
        long P, G, x, a, y, b, ka, kb;

        // Both the persons will agree upon the public keys G and P
        P = 23; // A prime number
        System.out.println("The value of P: " + P);

        G = 9; // A primitive root for P
        System.out.println("The value of G: " + G);

        // Alice will choose the private key a
        a = 4; // Alice's private key
        System.out.println("The private key a for Alice: " + a);

        x = power(G, a, P); // Alice's public key
        System.out.println("The public key x for Alice: " + x);

        // Bob will choose the private key b
        b = 3; // Bob's private key
        System.out.println("The private key b for Bob: " + b);

        y = power(G, b, P); // Bob's public key
        System.out.println("The public key y for Bob: " + y);

        // Generating the secret key after exchanging public keys
        ka = power(y, a, P); // Secret key for Alice
        kb = power(x, b, P); // Secret key for Bob
        System.out.println("Secret key for Alice is: " + ka);
        System.out.println("Secret key for Bob is:   " + kb);

        // Alice encrypts a message
        String message = "Hello Bob!";
        String encryptedMessage = encryptDecrypt(message, ka);
        System.out.println("Encrypted Message: " + encryptedMessage);

        // Bob decrypts the message
        String decryptedMessage = encryptDecrypt(encryptedMessage, kb);
        System.out.println("Decrypted Message: " + decryptedMessage);
    }
}

File: 04.2_diffie_hellman/diffie_hellman.py
# To run this code, open a terminal in this folder and run: python3 diffie_hellman.py

def power(a, b, P):
    """
    Calculates (a^b) % P using modular exponentiation.
    Python's built-in pow(a, b, P) is efficient for this.
    """
    return pow(a, b, P)

def encrypt_decrypt(message, key):
    """
    Encrypts or decrypts a message using a simple XOR cipher.
    Applying the same function twice with the same key returns the original message.
    """
    # ''.join() is an efficient way to build the resulting string
    return ''.join(chr(ord(char) ^ key) for char in message)

def main():
    """
    Main function to demonstrate the Diffie-Hellman key exchange.
    """
    # Both parties agree upon the public keys G and P.

    # A prime number P is taken.
    P = 23
    print(f"The value of P: {P}")

    # A primitive root for P, G is taken.
    G = 9
    print(f"The value of G: {G}")
    print("-------------------------")

    # Alice chooses her private key 'a'.
    a = 4
    print(f"The private key 'a' for Alice: {a}")

    # Alice calculates her public key 'x'.
    x = power(G, a, P)
    print(f"The public key 'x' for Alice: {x}")
    print("-------------------------")

    # Bob chooses his private key 'b'.
    b = 3
    print(f"The private key 'b' for Bob: {b}")

    # Bob calculates his public key 'y'.
    y = power(G, b, P)
    print(f"The public key 'y' for Bob: {y}")
    print("-------------------------")

    # After exchanging public keys, they generate the shared secret key.

    # Alice calculates the secret key.
    ka = power(y, a, P)

    # Bob calculates the secret key.
    kb = power(x, b, P)

    print(f"Secret key for Alice is: {ka}")
    print(f"Secret key for Bob is: {kb}")
    print("-------------------------")

    # Now they can use the shared secret key for encryption.
    message = "Hello Bob!"
    print(f"Original Message: {message}")

    # Alice encrypts the message with her key.
    encrypted_message = encrypt_decrypt(message, ka)
    print(f"Encrypted Message: {encrypted_message}")

    # Bob decrypts the message with his key.
    decrypted_message = encrypt_decrypt(encrypted_message, kb)
    print(f"Decrypted Message: {decrypted_message}")

# This ensures the main function is called only when the script is executed directly
if __name__ == '__main__':
    main()